### ISSUE TITLE
[object ArrayBufferConstructor] is not a function

### ISSUE THREAD

**Issue by SemihGk** [google/shaka-player#1022]
**Have you read the [FAQ](https://goo.gl/JE1Sy5) and checked for duplicate issues**:
Yes
**What version of Shaka Player are you using**:
2.2.0
**Can you reproduce the issue with our latest release version**:
Yes
**Can you reproduce the issue with the latest code from `master`**:
Yes
**Are you using the demo app or your own custom app**:
Custom app.
**If custom app, can you reproduce the issue using our demo app**:
No.
**What browser and OS are you using**:
Samsung 2016 Tv.  User-agent: Mozilla/5.0 (SMART-TV; Linux; Tizen 2.4.0) AppleWebKit/538.1 (KHTML, like Gecko) Version/2.4.0 TV

**What are the manifest and license server URIs**:
*(you can send the URIs to <shaka-player-issues@google.com> instead, but please use GitHub and the template for the rest)*
It only occurs in Samsung 2016 TV. If you have 2016 Samsung TV to test(also you would need Tizen Studio), I would be happy to share it. Otherwise, it is not reproducible.

**What did you do?**
When I try to load manifest into shaka.Player.load function, it throws this error.

**What did you expect to happen?**
It should not throw the error.

**What actually happened?**
First of all, this application works on latest browsers and 2017 Samsung TV as well. This error is only caught in 2016 TV. So, it is hard to test it. Honestly, I would not even say this is a valid bug since 2016 TV has very old version of WebKit browser. However, I am stuck at this point and I am wondering do you have any idea what it may cause this error. I will appreciate any help. Please check the screenshot. You will notice it does not even show the error line. Also, yes I checked ArrayBuffer is supported by the webkit browser. Thank you.

<img width="757" alt="screen shot 2017-09-13 at 4 02 32 pm" src="https://user-images.githubusercontent.com/5814582/30398602-0b3d0884-989f-11e7-9ae7-086a39f99bc0.png">

**Comment by TheModMaker** [comment]
First, can you try with the [uncompiled library](https://shaka-player-demo.appspot.com/docs/api/tutorial-debugging.html)?  It would be helpful to see a stack trace of this, the stack trace in the screenshot doesn't look right (or in the library).  Also, how are you loading Shaka Player?  The top-level object doesn't look familiar, is it a wrapped error from some other library or the platform?  Is `com.quickplay.plugins.shakaPlayer` from your app (again doesn't look familiar)?

**Comment by SemihGk** [comment]
Thank you for your reply, @TheModMaker . Yes, you are right! Stack trace is appeared when I inject the uncompiled library. I was importing the compiled.debug.js (it seems it is a bad idea). 
Track race 

```
 "exports↵exports↵slice
↵readBytes@file:///opt/usr/apps/GojX41wt1P/res/wgt/lib/util/data_view_reader.js:206:42
↵parseNext@file:///opt/usr/apps/GojX41wt1P/res/wgt/lib/util/mp4_parser.js:185:35
↵parse@file:///opt/usr/apps/GojX41wt1P/res/wgt/lib/util/mp4_parser.js:141:19
↵parseInit@file:///opt/usr/apps/GojX41wt1P/res/wgt/lib/text/mp4_ttml_parser.js:56:15
↵file:///opt/usr/apps/GojX41wt1P/res/wgt/lib/text/text_engine.js:145:29
↵[native code]↵run↵↵flus
```

Also, yes it is a wrapper from our application. It doesn't override the functions. It just initializes the shaka-player. Now, I am able to trace the bug. After some investigation, I will share the result. Thanks a lot again.

<img width="1043" alt="screen shot 2017-09-13 at 6 06 07 pm" src="https://user-images.githubusercontent.com/5814582/30403252-c214ddf2-98ae-11e7-9f68-a5cf36c0fdda.png">

**Comment by SemihGk** [comment]
After I digged into this bug more, I found out webkit browser partially does not support ArrayBuffer.slice() method. Specifically, slice method exist, but only takes one paramater. Second parameter was causing this error. So, I had to override the ArrayBuffer.slice method based on Claudiu's answer from this link [ArrayBuffer.slice override](https://stackoverflow.com/questions/21440050/arraybuffer-prototype-slice-shim-for-ie)

This topic can be closed unless you want to add this support for old browsers. Thank you for your help again.

**Comment by TheModMaker** [comment]
If that is the only bug here, I think we can work around it.  It would be nice to be able to support this browser.

**Comment by SemihGk** [comment]
Well, I cannot say this is the only bug. There is also another bug which was reported [here](https://github.com/google/shaka-player/issues/935) . I had to override cloneObject function based on [cloneDeep](https://www.npmjs.com/package/lodash.clonedeep). I do not receive any error so far. I can keep this topic updated if you are interested. Thank you.

**Comment by SemihGk** [comment]
@TheModMaker, I have been facing another issue. This might be more likely a StackOverflow question, but I will really appreciate any help. Because I am really desperate at this point. 

As I said above, this application works on latest browsers and not working in this WebKit browser. Problem is that video is not playing and there is no error (also listened 'uncaught errors' in case of missing promise catch blocks). That's why I do not know how to troubleshoot.

So far, I realized video blob file is created and video source is loaded into video element as well. Please check screenshots.  The question is here is it possible that browser cannot detect the video codecs or video type somehow? Or what are the other possibilities that I can look for? Please accept my apologies for the unrelated question in this topic.

![image uploaded from ios 1](https://user-images.githubusercontent.com/5814582/30491045-ea958c20-9a09-11e7-871a-22bf2467ba2f.jpg)
<img width="883" alt="screen shot 2017-09-15 at 10 51 17 am" src="https://user-images.githubusercontent.com/5814582/30491048-ece58fb6-9a09-11e7-9601-22aa2f545c80.png">

**Comment by TheModMaker** [comment]
Here are some things to check:  
* Are we downloading segments?
* `video.readyState`, `video.playbackRate`,  `video.paused`, `video.error`
* `player.isBuffering()` (in the default demo it would be `shakaDemo.localPlayer_.isBuffering()`)
* video events: `playing`, `loadedmetadata`, `waiting` (add listener before calling `player.load()`)
* Try calling `video.play()` and try setting `video.currentTime` to force a seek

What is the value of `video.buffered`?  You'll need to manually get the ranges, for example:
```js
var b = [];
for (var i = 0; i < video.buffered.length; i++)
  b.push([video.buffered.start(i), video.buffered.end(i)]);
```

If you are using the uncompiled library, you can also set the log level to get more info.  This could be used to see if something stalls or if we append segments successfully.

```js
shaka.log.setLevel(shaka.log.Level.DEBUG);
shaka.log.setLevel(shaka.log.Level.V1);
shaka.log.setLevel(shaka.log.Level.V2);
```

**Comment by SemihGk** [comment]
@TheModMaker. Thank you so much for the reply. Actually, I checked some of them before, but it does not still explain what's the problem for me.

- Are we downloading segments?
Yes, I checked the manifest. It has the segmets.
- video.readyState, video.playbackRate, video.paused, video.error
when loadedmetadata is fired: readyState: 3, playbackRate: 0, paused: true, error: null, 
if I try to force to play, readyState: 0, playbackRate: 0, paused: false, error: MediaError code 4.
- player.isBuffering() (in the default demo it would be shakaDemo.localPlayer_.isBuffering())
returning false
- video events: playing, loadedmetadata, waiting (add listener before calling player.load())
only loadedmetadata is triggered.
- Try calling video.play() and try setting video.currentTime to force a seek
Tried it. Nothing happens. Also, there is no error.
However, when I tried video.play() and video.pause(), an error in video element appears which is MediaError code 4(please check screenshots) although shaka does not throw an error.

- video.buffered: 0 . So, not possible to get ranges.

Do you have any idea why buffered length is 0? Because, I checked the working version. That one's buffered length is 1. What may it cause this? Or what else can I check? 

**When its first loaded:** 

<img width="315" alt="screen shot 2017-09-18 at 12 32 00 pm" src="https://user-images.githubusercontent.com/5814582/30552858-6d789686-9c6d-11e7-8646-05bde0efd788.png">
<img width="325" alt="screen shot 2017-09-18 at 12 31 44 pm" src="https://user-images.githubusercontent.com/5814582/30552860-6d7bf6fa-9c6d-11e7-94a7-7437c01ca2ab.png">

**After forced to play and pause:** 

<img width="560" alt="screen shot 2017-09-18 at 12 31 34 pm" src="https://user-images.githubusercontent.com/5814582/30552859-6d7bb794-9c6d-11e7-8469-38cd07015bfc.png">
<img width="555" alt="screen shot 2017-09-18 at 12 31 17 pm" src="https://user-images.githubusercontent.com/5814582/30552861-6d7c73b4-9c6d-11e7-88e7-844e25d07330.png">

**Comment by TheModMaker** [comment]
> * Are we downloading segments?
> Yes, I checked the manifest. It has the segmets.

I was actually referring to checking the network tab to see if we are actually downloading them.

> Do you have any idea why buffered length is 0? Because, I checked the working version. That one's buffered length is 1. What may it cause this? Or what else can I check?

A length of 0 means no content is buffered.  This can happen before we download segments, or if the content is not supported.

Media error 4 is [MEDIA_ERR_SRC_NOT_SUPPORTED](https://dev.w3.org/html5/spec-author-view/video.html#error-codes).  It looks like the browser doesn't like the MSE source.  I see there is a `mediaKeys` object that doesn't look correct.  Are you able to play clear content?  Can you try other manifests to see if this is a content problem?

**Comment by SemihGk** [comment]
> I was actually referring to checking the network tab to see if we are actually downloading them.

Oh, sorry. That's my bad. You are right. It does not download all contents. It stops downloading after 5 chunks (see screenshot). The working one keeps downloading each segment while its playing. However, it does not make sense to me if it is content is unavailable issue. Because, same urls are requested between Chrome one and TV browser one. It should download all contents. Plus, this problems are exists at all different contents (live or VOD). 

> A length of 0 means no content is buffered. This can happen before we download segments, or if the content is not supported.
Could you expand "the content is not supported" issue? Does it mean content is simply unavailable(which is unlikely since I can play the content) or content may not be encrypted in the browser?

> Media error 4 is MEDIA_ERR_SRC_NOT_SUPPORTED. It looks like the browser doesn't like the MSE source. I see there is a mediaKeys object that doesn't look correct. Are you able to play clear content? Can you try other manifests to see if this is a content problem?
MediaKeys object is different than chrome. I realized that as well, then I thought this must be added by samsung since webkit(safari) does not support playready natively. Yes, I am able to play clear contents, but I cannot play widevine or playready contents. Regarding this topic [771](https://github.com/google/shaka-player/issues/771) , I am having the same issue for the demo app as well. I cannot play any content except first one in the demo app. I see samsung documentation has the playready DRM [playready](http://developer.samsung.com/tv/develop/legacy-platform-library/tut00064/index) . Do you think this issue may be an encryption issue? Since I am new at shaka-player, I do not know how to trace it from this point if this is the issue. 

Thank you so much for your help.

<img width="998" alt="screen shot 2017-09-18 at 3 20 29 pm" src="https://user-images.githubusercontent.com/5814582/30560446-f6013df6-9c85-11e7-834a-4fa7d874a1d4.png">

**Comment by TheModMaker** [comment]
From issue #771 it looks like Tizen 2.4 doesn't implement the APIs we need to use PlayReady.  The documentation you linked to shows that it uses a custom API to use PlayReady, which we don't support.  So it looks like that browser isn't supported for protected content, you can follow #771 to get updates.

**Comment by SemihGk** [comment]
Hmm, I never thought playready is supported in their custom API. I will follow the #771 . However, I will ask Samsung about playready support. Thank you!

**Comment by joeyparrish** [comment]
The first part of the fix has been cherry-picked to v2.2.x and will be released in v2.2.2.  A deeper refactor is coming, which will land in v2.3.0 due to small API changes.
